### PR TITLE
7529 do not disable autocast for cuda devices

### LINKED ISSUE #7529: ValueError: For the given accelerator, there seems to be an unexpected problem in type-casting. Please file an issue on the PyTorch GitHub repository.
### Describe the bug

Running the script examples/text_to_image/train_text_to_image_lora_sdxl.py
Not working anymore on mixed_precision fp16, bf16

Traceback (most recent call last):0%|██████████████████████████████████████████████████████████████| 7/7 [00:00<00:00, 67.30it/s]
  File "train_text_to_image_lora_sdxl.py", line 1350, in <module>
    main(args)
  File "train_text_to_image_lora_sdxl.py", line 1229, in main
    images = [
  File "train_text_to_image_lora_sdxl.py", line 1230, in <listcomp>
    pipeline(**pipeline_args, generator=generator).images[0]
  File "/home/ubuntu/360-training-preprocess/venv/lib/python3.8/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/ubuntu/360-training-preprocess/venv/lib/python3.8/site-packages/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py", line 1203, in __call__
    raise ValueError(
ValueError: For the given accelerator, there seems to be an unexpected problem in type-casting. Please file an issue on the PyTorch GitHub repository. See also: https://github.com/huggingface/diffusers/pull/7446/.

### Reproduction

To me, the regression comes from recent commit https://github.com/huggingface/diffusers/commit/544710ef0fe83c3ea87744c21a550d46e088a3c4

https://github.com/huggingface/diffusers/pull/7446

I've never experiment this issue before. 

Of course latents are in mixed precision before, and they are in float32 after passing through scheduler.

### Logs

_No response_

### System Info

Ubuntu, A10G,  NVIDIA-SMI 550.54.14, Driver Version: 550.54.14, CUDA Version: 12.4
diffusers==0.28.0.dev0

### Who can help?

@bghira @sayakpaul

### PR DESCRIPTION
**Pull request by bghira**
# What does this PR do?

Fixes #7529

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by sayakpaul — APPROVED** [review]
Thanks for taking care of it!

**Comment by bghira** [comment]
i'm going to switch from this method to contextlib's nullcontext, as local testing with nightly pytorch indicates that'll avoid some new errors that will come down the pipe, and specifically the null context will allow us to entirely bypass all issues with autocast - sometimes, the platform is disabled incorrectly, or has partial support.

the nullcontext will allow us to fully decide when to commit to providing autocast on a new platform

**Comment by bghira** [comment]
@sayakpaul ready :-)

**Comment by bghira** [comment]
@pcuenca i tried to unify the behaviour of the context selection to use contextlib

**Review by sayakpaul — APPROVED** [review]
Nice 👌

**Comment by simbrams** [comment]
Can you also remove the error test in the instruct pix2pix xl pipeline 🙏 : https://github.com/huggingface/diffusers/blob/5266ab7935dd9e9aec596cdc2464badf1eacd99a/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_instruct_pix2pix.py#L927-L930

**Comment by bghira** [comment]
@simbrams done

**Comment by bghira** [comment]
@sayakpaul @DN6 @yiyixuxu i've updated every pipeline that uses 🤗 Accelerate to disable AMP so that `accelerator.prepare(...)` does not try loading an autocast ctx.

**Review by christopher5106 — APPROVED** [review]
Works for me now. I mainly checked the pipeline and the train script for sdxl lora.

**Comment by sayakpaul** [comment]
@bghira thanks for the changes. A couple of comments:

* GPUs are first-class citizens for the training scripts. In order to support Silicon compatibility, we cannot break that. I am sorry that I didn't check that earlier.
* I see we're now making changes to the pipelines so that they use an inference context. I don't think we can have that level of change in our pipelines. The silicon support is experimental from PyTorch itself. So when that gets stabilized a bit, we might have to undo the changes we're currently making. So, I am not in favor of going via this route. 
* IMO, the changes should be kept as minimal as possible. If making changes to the training scripts and the schedulers (like we did before) to support Silicon compatibility isn't enough, that means this feature itself is a little too experimental to support as that requires non-trivial changes to the legacy core. 

So, IMO, we should make sure:

* Training works as before on GPUs. 
* Silicon training works but with minimal changes. If that prevents FP16 training, I would still be okay with that because it seems like it's quite experimental. If disabling the intermediate validation could work -- that could be another route too (albeit not a pragmatic one).

@pcuenca @yiyixuxu would like to get your opinions too.

**Comment by yiyixuxu** [comment]
@sayakpaul 

> I see we're now making changes to the pipelines so that they use an inference context.

I didn't see it in this PR though

**Comment by yiyixuxu** [comment]
@sayakpaul 
>So, IMO, we should make sure:

>Training works as before on GPUs.
>Silicon training works but with minimal changes. If that prevents FP16 training, I would still be okay with that because it >seems like it's quite experimental. If disabling the intermediate validation could work -- that could be another route too (albeit not a pragmatic one).


agree with this

**Comment by sayakpaul** [comment]
> I didn't see it in this PR though

@yiyixuxu see here: https://github.com/huggingface/diffusers/pull/7530/files#diff-942ed0e4f11fae99750e9b042c4950baa2078bf7f75b1a450a7fa0339f5f034bR551

image

**Comment by bghira** [comment]
that was already using a context manager, i simply updated it to be consistent

**Comment by sayakpaul** [comment]
Apologies for the oversight. You're right.

**Comment by bghira** [comment]
i was able to reproduce the MPS crash from before, but it's an issue with the *model weights*. using the single file non-ema pruned checkpoint for SD 2.1-v even runs in fp16 mode without NaNs, but using the `stabilityai/stable-diffusion-2-1` weights as-is will result in NaNs on the very first step of inference. i checked where it comes from, and it's nn.Linear, but this occurs on CPU, CUDA, AMD, and Apple.

when we use 2.1-base, it does not cause this NaN issue, and everything is fine. the non-ema pruned 2.1-v checkpoint is also okay. it seems like there is something wrong with the extracted diffusers weights for 2.1-v. i have no expertise with this however

**Comment by bghira** [comment]
as an aside, i've finetuned the issue out of 2.1-v weights. so, a fixed model is available ... but it seems like the actual 2.1-v weights should be fixable somehow with a new or fixed extraction.

**Comment by sayakpaul** [comment]
Okay good to know. Thanks!

What are our blockers now? I am happy to cook up a context manager in `training_utils.py` and then we can take it from there?

**Comment by bghira** [comment]
also, the instruct pix2pix script is where a lot of the original confusion came from! that you wrote :D there, the autocast disables for fp16 and pulls the device type str to chop `:0` off - so i learned a few things along the way to fixing this. but now everything is at least consistent - there were some community pipelines i did not update, as it didn't seem appropriate to do so.

**Comment by bghira** [comment]
i've been trying to chase down that training issue. i want to try running SD 1.5 and see if the problem for legacy models is limited to 2.x, in which case it'll be best to limit training to SD 1.x.. will update shortly

**Comment by sayakpaul** [comment]
> the autocast disables for fp16 and pulls the device type str to chop :0 off 

I see it's enabled when mixed precision is FP16:

https://github.com/huggingface/diffusers/blob/7956c36aaa2b43584f16f7e755e46aa2d7efa721/examples/instruct_pix2pix/train_instruct_pix2pix.py#L947

But anyway, as mentioned I am down to writing a context manager to help ease things a bit. Maybe providing everything you have discovered so far regarding training on Silicon could be a very nice resource.

**Comment by bghira** [comment]
on mps,
* sd 2.1 does not work in diffusers or simpletuner
* sd 1.5 does not work in diffusers or simpletuner
  * both of these have MPS internal crashes
* sdxl does not work in diffusers, but does work in simpletuner
  * in diffusers, fp16 training on sdxl causes the attention processor query to be in float32 while the weights are half

i've gone through them a lot today, and really not sure why it would cause the MPS crash, it's incredibly unfortunate.

the SDXL error does seem like it can be overcome, i just need to put a bit more time into hunting it down

so far, this patch works as advertised at least to resolve the cuda-specific regression and it should probably be merged

**Comment by sayakpaul** [comment]
That's quite informative indeed. 

> sdxl does not work in diffusers, but does work in simpletuner
in diffusers, fp16 training on sdxl causes the attention processor query to be in float32 while the weights are half

Curious, what do you do on `simpletuner` to have that fixed?

**Comment by sayakpaul** [comment]
I will wait for @pcuenca to give it a shot too, before merging.

**Comment by bghira** [comment]
in simpletuner i have a lot more brute-force general handling of dtypes and leave less responsibility for that on other parts of the stack.

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py
Comment (DN6) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:2195]: Autocast isn't supported for Playground models?

Comment (sayakpaul) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:2195]: Nope. The validation runs go crazy when doing autocasting for the Playground model. More details: https://github.com/huggingface/diffusers/pull/7126

**Comment by bghira** [comment]
good morning. after some more checking, fp32 training works fine for MPS on SD 1.5 and 2.1. but this feels disappointing, as dtype issues just shouldn't be the only thing holding back memory-efficient training. but i guess it's all we can do for now?

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py
Comment (bghira) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:2195]: autocast in general seems like a huge mistake to rely on.

last year the advice from patrick von platen was to never use it for inference, and here we are, abusing the heck out of it at inference time

Comment (sayakpaul) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:2195]: Fair to be skeptical. But I am really hesitant to touch what's been working for the majority of our users (GPU training i.e.,) for a while now. 

>  abusing the heck out of it at inference time

Little err. Only during intermediate validation runs during training.

Comment (bghira) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:2195]: did you know autocast when running training in bf16, will ruin every tensor by casting them down to fp16 and then back up to bf16? :D

Comment (sayakpaul) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:2195]: Nope. Cannot possibly exhaustively know everything beforehand.

Comment (AmericanPresidentJimmyCarter) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:2195]: > Nope. Cannot possibly exhaustively know everything beforehand.

Something very bad is happening with autocast. See my replies in this issue here: https://github.com/pytorch/pytorch/issues/120376

Basically bfloat16 performance should be matching float32 and I rewrote the AdamW optimiser to do so, but you can see in the graph that when you store the weights in float32 and then autocast to bfloat16 training completely breaks.

I believe this problem also affects all accelerate scripts that commonly store weights in float32 and then train in bfloat16. And with the improved optimiser, everyone should really just be storing weights in bfloat16 and training with bfloat16 autocast anyway. We really need to get torch in on this to fix the autocast feature, because something appears to be going very wrong.

Comment (bghira) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:2195]: i am in agreement. i had autocast disabled in simpletuner for about a year and now that it's back in with my changes this week, things have become a nightmare for reliability

File: examples/controlnet/train_controlnet.py
Comment (sayakpaul) [inline examples/controlnet/train_controlnet.py:755]: @bghira possible to add a more descriptive comment here?

**Comment by bghira** [comment]
the fp16 mps bug happens on the third call to nn.Linear, the Parameter object is created with fp32 dtype for weight and bias but the query is fp16

**Comment by bghira** [comment]
its an actual training failure on step 1 when t=999

i saw that nn.Linear has a dtype parameter i tried setting but the Parameter class it gets passed into actually doesnt make use of the dtype parameter which seems like a torch bug

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py
Comment (AmericanPresidentJimmyCarter) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:2195]: Kenji on the LAION Discord was able to figure out how to fix autocast with float32 weights and bfloat16, basically autocast is dangerous to use outside of the loss calculate and the `forward` step as it will completely corrupt the optimisation step. Once you put it there (and ONLY there) the optimiser will work in float32 and everything should be fixed.

Screenshot_2024-04-02_09-22-08

That said if we have a good bf16 version of AdamW now, it makes sense to save on the weights and just use that.

**Comment by sayakpaul** [comment]
Thanks very much folks for the insightful discussions on autocast and bfloat16. I am under the impression that PyTorch, indeed, has some weird voodoo going under the hood for these. 

Meanwhile, I am going to merge this PR to unblock our users. So, thanks to @bghira for the prompt action here. 

This thread still remains open for discussions around autocast and bfloat16 which I believe will be valuable for the community.

**Comment by bghira** [comment]
without attention slicing, we see:

```
/AppleInternal/Library/BuildRoots/ce725a5f-c761-11ee-a4ec-b6ef2fd8d87b/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:788: failed assertion `[MPSNDArray initWithDevice:descriptor:] Error: total bytes of NDArray > 2**32'
```

but with attention slicing, this error disappears.

the resolution seems to be:

```py
    # Base components to prepare
    if torch.backends.mps.is_available():
        accelerator.native_amp = False
    results = accelerator.prepare(unet, lr_scheduler, optimizer, *train_dataloaders)
    unet = results[0]
    if torch.backends.mps.is_available():
        unet.set_attention_slice()
```

however, the casting type issue is now there for bf16 **and** fp16, but bf16 no longer crashes. still digging into it.

**Comment by AmericanPresidentJimmyCarter** [comment]
> Thanks very much folks for the insightful discussions on autocast and bfloat16. I am under the impression that PyTorch, indeed, has some weird voodoo going under the hood for these.
> 
> Meanwhile, I am going to merge this PR to unblock our users. So, thanks to @bghira for the prompt action here.
> 
> This thread still remains open for discussions around autocast and bfloat16 which I believe will be valuable for the community.

The issues seem to be scoped to bfloat16, which is unfortunate because most hardware has moved to it over the past several years.

>     The assert_close should fail. Interestingly, the assert_close passes if we change the autocast dtype to torch.float16.
https://github.com/pytorch/pytorch/issues/120930
>     A model with PyTorch AMP produces different model results when used in conjunction with CUDA graphs.
    In other words, it seems that conducting CUDA graph capture while wrapped with autocast leads to wrong outputs during training (specifically, after the first weight update).
https://github.com/pytorch/pytorch/issues/71631

The second issue is from Jan 2022, so we may still be waiting for a while for a fix. It is clear that autocast was originally designed with fp16 and the fp16 grad scaler in mind and has issues with other types.

**Comment by bghira** [comment]
so disabling autocast on bf16 everywhere does seem like the way to go. honestly the new optimiser solves any need to use fp16 - pure bf16 is lighter weight and simpler code

**Pull request merged**
